packages/material-ui/src/ClickAwayListener/ClickAwayListener.test.js

```diff
@@ -185,11 +175,80 @@ describe('<ClickAwayListener />', () => {
         }
         render(<Test />);
 
-        fireDiscreteEvent.click(screen.getByTestId('trigger'));
+        const fireEventName = `${mouseEvent[2].toLowerCase()}${mouseEvent.slice(3)}`;
+        fireDiscreteEvent[fireEventName](screen.getByTestId('trigger'));
 
         expect(screen.getByTestId('child')).not.to.equal(null);
       });
     });
+
+    it('should be called if an element is interleaved between mousedown and mouseup', () => {
+      /**
+       * @param {Element} element
+       * @returns {Element[]}
+       */
+      function ancestorElements(element) {
+        const ancestors = [];
+        let ancestor = element;
+        while (ancestor !== null) {
+          ancestors.unshift(ancestor);
+          ancestor = ancestor.parentElement;
+        }
+        return ancestors;
+      }
+
+      /**
+       * @param {Element} elementA
+       * @param {Element} elementB
+       * @returns {Element}
+       */
+      function findNearestCommonAncestor(elementA, elementB) {
+        const ancestorsA = ancestorElements(elementA);
+        const ancestorsB = ancestorElements(elementB);
+
+        if (ancestorsA[0] !== ancestorsB[0]) {
+          throw new Error('A and B share no common ancestor');
+        }
+
+        for (let index = 1; index < ancestorsA.length; index += 1) {
+          if (ancestorsA[index] !== ancestorsB[index]) {
+            return ancestorsA[index - 1];
+          }
+        }
+
+        throw new Error('Unreachable reached. This is a bug in findNearestCommonAncestor');
+      }
+
+      const onClickAway = spy();
+      function ClickAwayListenerMouseDownPortal() {
+        const [open, toggleOpen] = React.useReducer((flag) => !flag, false);
+
+        return (
+          <ClickAwayListener onClickAway={onClickAway}>
+            <div data-testid="trigger" onMouseDown={toggleOpen}>
+              {open &&
+                // interleave an element during mousedown so that the following mouseup would not be targetted at the mousedown target.
+                // This results in the click event being targetted at the nearest common ancestor.
+                ReactDOM.createPortal(
+                  <div data-testid="interleaved-element">Portaled Div</div>,
+                  document.body,
+                )}
+            </div>
+          </ClickAwayListener>
+        );
+      }
+      render(<ClickAwayListenerMouseDownPortal />);
+      const mouseDownTarget = screen.getByTestId('trigger');
+
+      fireDiscreteEvent.mouseDown(mouseDownTarget);
+      const mouseUpTarget = screen.getByTestId('interleaved-element');
+      // https://w3c.github.io/uievents/#events-mouseevent-event-order
+      const clickTarget = findNearestCommonAncestor(mouseDownTarget, mouseUpTarget);
+      fireDiscreteEvent.mouseUp(mouseUpTarget);
+      fireDiscreteEvent.click(clickTarget);
+
+      expect(onClickAway.callCount).to.equal(1);
```

Comment:
I don't buy this test case. I see it failing in two dimensions.

First, at the product level, I think that developer should expect 

```suggestion
      expect(onClickAway.callCount).to.equal(0);
```

Second, at the technical level, I can revert the change in the source to HEAD, and apply this diff:

```diff
       /**
        * @param {Element} element
        * @returns {Element[]}
@@ -238,6 +238,7 @@ describe('<ClickAwayListener />', () => {
         );
       }
       render(<ClickAwayListenerMouseDownPortal />);
+      clock.tick(0); // reproduce real-life conditions.
       const mouseDownTarget = screen.getByTestId('trigger');

       fireDiscreteEvent.mouseDown(mouseDownTarget);
```

and get it green, so I don't think that we test the behavior.

Reply:
I think I broke something during rebasing. The goal of this PR was to get it to 0.

> clock.tick(0); // reproduce real-life conditions.

That's not necessarily real life conditions.

Reply:
> That's not necessarily real life conditions.

True, maybe only more frequent? 👼

Reply:
Went over this again, and this is expected behavior for now. If we want to change that behavior then we need to rework the implementation to our new notion of "click-away". In this new behavior we don't mean "click" in how the DOM specifies it. Rather we want to consider the mouseup target. So we'd probably just introduce a "mouse-press" where we only look at the mouseup target. Though we'd still dispatch it either during the click phase or the configured `mouseEvent` phase. Whichever is earlier.

Though I'd be very careful with doing this inside the ClickAwayListener and rather introduce a new, experimental components. It's not at all guaranteed that the new behavior is strictly better behavior. There's probably a reason why `click` is targetted the way it is.

Reply:
Adding the test in a separate PR. The test is relevant regardless of this PR.